### PR TITLE
Create option for disabling strong name check during serialization

### LINKED ISSUE #3036: Consider adding an option for MobileFormatter to ignore version/signing on deserialization
Although very risky and unsafe for a user, it may be possible for `MobileFormatter` to have an option to allow deserialization into a type of the same name/assembly where the version and signing information doesn't match.

The referenced discussion thread contains info about why this is risky and may lead to people having hard-to-find and hard-to-solve issues in their app.

---

Number 4: Apparently we are not using the SerializationInfo correctly. We add all of an object's properties using info.AddValue function regardless of if the value is a custom type or IMobileObject type or System type. 

1. Add an option to turn on assembly/type matching somewhere in the CSLA options (default setting is off)
2. Change MobileFormatter to look at the matching options

Should just be those two.

_Originally posted by @crazyfox55 in https://github.com/MarimerLLC/csla/issues/3021#issuecomment-1196007880_

### PR DESCRIPTION
**Pull request by kant2002**
Closes #3036

### CODE REVIEW

File: Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs [hunk 1]
```diff
@@ -18,6 +18,11 @@ public class MobileFormatterOptions
   /// Gets the list of custom serializers.
   /// </summary>
   public List<ITypeMap> CustomSerializers { get; } = [];
+  
+  /// <summary>
+  /// Gets a value indicating whether to use strong name checks during serialization/deserialization process.
+  /// </summary>
+  public bool UseStrongNamesCheck { get; internal set; } = true;
```
Comment (rockfordlhotka) [inline Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs:25]: Shouldn't this be a private set?

Or just make the setter public and let the developer set the property during configuration (skipping the two methods)?

**Review by rockfordlhotka — CHANGES_REQUESTED** [review]
There needs to be some tests for this.

At the very least, use reflection to invoke the method that returns the type name to ensure it does and does not include the version info.

File: Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs (hunk 1 above)
Comment (kant2002) [reply Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs:25]: I have impression that you want fluent interface for the options. At least this is seems to be the pattern. Maybe it make sense to have both imperative and fluent API?

**Comment by StefanOssendorf** [comment]
I prefer the fluent interface. And I have a task (in my personal list) to
make all options fluents when I have the time xD

Andrii Kurdiumov ***@***.***> schrieb am Do., 11. Juli 2024,
07:26:

File: Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs (hunk 1 above)
Comment (rockfordlhotka) [reply Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs]: True, fluent is easier for the end developer. The setter should still be private right?

Also, the fluent method could accept a parameter:

```c#
  public MobileFormatterOptions StrongNamesCheck(bool check)
    => UseStrongNamesCheck = check;
```

Comment (kant2002) [reply Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs]: I prefer to have it both ways, that's most inclusive in my opinion. that way library do not force preferences on developers, and I think this is not super complicated to maintain in the long run.

Comment (kant2002) [reply Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs]: I have mixed signals. Error! Error! Operator intervention required.

Comment (rockfordlhotka) [reply Source/Csla/Serialization/Mobile/Configuration/MobileFormatterOptions.cs]: Let's go with Stefan's model, I see the value in the argument.

**Review by StefanOssendorf — APPROVED** [review]
Looks good to me. lgtm @rockfordlhotka

**Review by rockfordlhotka — APPROVED** [review]

**Review by rockfordlhotka — APPROVED** [review]

**Pull request merged**